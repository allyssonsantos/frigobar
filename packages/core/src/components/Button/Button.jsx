import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Icon from '@material-ui/core/Icon';

const StyledIcon = styled(Icon)`
  margin-right: ${({ size }) => (size === 'large' ? '8px' : '6px')};
`;

const Btn = styled.button`
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 0.875rem;
  transition: all 200ms ease-in-out;

  ${({
    theme: {
      components: {
        button: { backgroundColor, textColor, sizes },
      },
    },
    skin,
    icon,
    size,
    disabled,
    full,
  }) => `
    background-color: ${backgroundColor[skin]};
    color: ${textColor.enabled};
    font-size: ${sizes[size].font}rem;
    padding: ${sizes[size].padding.top}px ${sizes[size].padding.right}px ${
    sizes[size].padding.bottom
  }px ${sizes[size].padding.left}px;

    ${
      disabled
        ? `
      background-color: ${backgroundColor.disabled};
      color: ${textColor.disabled};
      cursor: not-allowed;
    `
        : `
        &:hover {
          box-shadow: 0px 1px 3px 0px rgba(0, 0, 0, 0.4);
        }
        `
    }

    ${StyledIcon} {
      font-size: ${sizes[size].icon}rem;
    }

    ${
      icon
        ? `
      display: flex;
      align-items: center;
      justify-content: center;
    `
        : ''
    }

    ${full ? 'width: 100%;' : ''}
  `}
`;

const Button = ({ children, icon, theme, size, full, ...props }) => (
  <Btn icon={icon} theme={theme} size={size} full={full} {...props}>
    {icon && <StyledIcon size={size}>{icon}</StyledIcon>}
    {children}
  </Btn>
);

Button.propTypes = {
  children: PropTypes.node.isRequired,
  skin: PropTypes.oneOf(['primary', 'info', 'success', 'danger', 'warning', 'neutral']),
  disabled: PropTypes.bool,
  icon: PropTypes.string,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  full: PropTypes.bool,
};

Button.defaultProps = {
  skin: 'neutral',
  disabled: false,
  icon: undefined,
  size: 'small',
  full: false,
};

export default Button;
