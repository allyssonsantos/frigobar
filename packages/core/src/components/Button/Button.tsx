import React from 'react';
import styled from 'styled-components';

interface ButtonProps {
  children: React.ReactNode;
  /** Change background-color */
  skin?: 'primary' | 'info' | 'success' | 'danger' | 'warning' | 'neutral';
  /** change the width, height and font-size values */
  size?: 'small' | 'medium' | 'large' | undefined;
  /** Change disabled state of the button */
  disabled?: boolean;
  /** 100% width */
  full?: boolean;
  /** add border-radius circle */
  rounded?: boolean;
  /** large button */
  large?: boolean;
}

const Button = styled.button<ButtonProps>`
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 0.875rem;
  transition: all 200ms ease-in-out;
  text-decoration: none;

  ${({
    theme: {
      radius,
      components: {
        button: { backgroundColor, textColor, sizes },
      },
    },
    skin,
    size,
    disabled,
    full,
    rounded,
    large,
  }) => `
    background-color: ${backgroundColor[skin]};
    color: ${textColor.enabled};
    font-size: ${large ? '1.125' : sizes[size].font}rem;
    padding: ${
      large
        ? `${sizes[size].padding.top * 2}px
           ${sizes[size].padding.right * 2}px
           ${sizes[size].padding.bottom * 2}px
           ${sizes[size].padding.left * 2}px
        `
        : `${sizes[size].padding.top}px
           ${sizes[size].padding.right}px
           ${sizes[size].padding.bottom}px
           ${sizes[size].padding.left}px
        `
    };

    ${
      disabled
        ? `
      background-color: ${backgroundColor.disabled};
      color: ${textColor.disabled};
      cursor: not-allowed;
    `
        : `
        &:hover {
          box-shadow: 0px 1px 3px 0px rgba(0, 0, 0, 0.4);
        }
        `
    }

    ${full ? 'width: 100%;' : ''}
    ${rounded ? `border-radius: ${radius[3]}px` : ''}
  `}
`;

Button.defaultProps = {
  skin: 'neutral',
  disabled: false,
  size: 'small',
  full: undefined,
  rounded: undefined,
  large: undefined,
};

export default Button;
