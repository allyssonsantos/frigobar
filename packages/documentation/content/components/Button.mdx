---
title: Button
menu: component
---

Button component

<Button>Button</Button>
