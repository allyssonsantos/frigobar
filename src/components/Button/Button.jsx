import styled from 'styled-components';
import defaultTheme from '../../theme';

const Button = styled.button`
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 14px;
  padding: 6px 14px;
  transition: all 200ms ease-in-out;

  &:hover {
    box-shadow: 0px 1px 3px 0px rgba(0, 0, 0, 0.4);
  }

  ${({ theme, skin }) => `
    background-color: ${theme.button[skin].bgColor}
    color: ${theme.button[skin].textColor}
  `}
`;

Button.defaultProps = {
  theme: defaultTheme,
  skin: 'neutral',
};

export default Button;
