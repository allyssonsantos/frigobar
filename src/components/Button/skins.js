const getButtonColors = (colors, skin) => ({
  bgColor: colors[skin][500],
  textColor: colors.neutral[50],
});

const skins = ({ colors }) => ({
  primary: getButtonColors(colors, 'primary'),
  info: getButtonColors(colors, 'info'),
  success: getButtonColors(colors, 'success'),
  warning: getButtonColors(colors, 'warning'),
  danger: getButtonColors(colors, 'danger'),
  neutral: getButtonColors(colors, 'neutral'),
});

export default skins;
