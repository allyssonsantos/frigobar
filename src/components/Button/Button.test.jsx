import React from 'react';
import { render } from '@testing-library/react';

import Button from './Button';

describe('<Button />', () => {
  describe('Snapshots', () => {
    it('should match snapshot with all skins', () => {
      const buttons = [
        <Button />,
        <Button skin="primary" />,
        <Button skin="info" />,
        <Button skin="success" />,
        <Button skin="warning" />,
        <Button skin="danger" />,
      ];

      buttons.forEach(Component => {
        const { container } = render(Component);
        expect(container).toMatchSnapshot();
      });
    });
  });
});
