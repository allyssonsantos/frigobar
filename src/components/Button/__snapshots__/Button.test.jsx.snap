// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Snapshots should match snapshot with all skins 1`] = `
.c0 {
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 14px;
  padding: 6px 14px;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
  background-color: #999999;
  color: #ffffff;
}

.c0:hover {
  box-shadow: 0px 1px 3px 0px rgba(0,0,0,0.4);
}

<div>
  <button
    class="c0"
  />
</div>
`;

exports[`<Button /> Snapshots should match snapshot with all skins 2`] = `
.c0 {
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 14px;
  padding: 6px 14px;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
  background-color: #598aed;
  color: #ffffff;
}

.c0:hover {
  box-shadow: 0px 1px 3px 0px rgba(0,0,0,0.4);
}

<div>
  <button
    class="c0"
  />
</div>
`;

exports[`<Button /> Snapshots should match snapshot with all skins 3`] = `
.c0 {
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 14px;
  padding: 6px 14px;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
  background-color: #25387f;
  color: #ffffff;
}

.c0:hover {
  box-shadow: 0px 1px 3px 0px rgba(0,0,0,0.4);
}

<div>
  <button
    class="c0"
  />
</div>
`;

exports[`<Button /> Snapshots should match snapshot with all skins 4`] = `
.c0 {
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 14px;
  padding: 6px 14px;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
  background-color: #50a47f;
  color: #ffffff;
}

.c0:hover {
  box-shadow: 0px 1px 3px 0px rgba(0,0,0,0.4);
}

<div>
  <button
    class="c0"
  />
</div>
`;

exports[`<Button /> Snapshots should match snapshot with all skins 5`] = `
.c0 {
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 14px;
  padding: 6px 14px;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
  background-color: #cd9447;
  color: #ffffff;
}

.c0:hover {
  box-shadow: 0px 1px 3px 0px rgba(0,0,0,0.4);
}

<div>
  <button
    class="c0"
  />
</div>
`;

exports[`<Button /> Snapshots should match snapshot with all skins 6`] = `
.c0 {
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 14px;
  padding: 6px 14px;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
  background-color: #f44336;
  color: #ffffff;
}

.c0:hover {
  box-shadow: 0px 1px 3px 0px rgba(0,0,0,0.4);
}

<div>
  <button
    class="c0"
  />
</div>
`;
